docs/historical/mh-jun-1982/man/news.1
.TH NEWS 1 RAND
.SH NAME
news \- show news
.SH SYNOPSIS
\fBnews\fR \%[topic\ ...]
\%[\fB\-check\fR]
\%[\fB\-display\fR]
\%[\fB\-update\fR]
\%[\fB\-review\fR\ [topic]]
\%[\fB\-send\ \fRtopic]
\%[\fB\-topics\fR]
\%[\fB\-help\fR]
.SH DESCRIPTION
.I News
is an MH program which checks, displays, reviews, or sends news.
A pseudo-user called `news' acts as an automated
custodian for news items on various topics.
These topics are managed as MH folders in the home directory of news.
Mail sent to `news.everyone', for instance, will get
filed in the `everyone' news folder.
.PP
You can explicitly select the group of news topics you want
to check or display by specifying them as arguments to the
.I news
command.
If you don't explicitly select topics in this way,
the topic `everyone' is selected by default.
Additional default topics can be selected by listing them on a
\&``News\-Topics:'' line in your MH profile.
The MH profile line
must contain a blank-separated list of topics.
As a special case, the profile line:
.br
    \&``News\-Topics:\ *''
.br
means you want to keep track of all news.
.PP
\&``\fINews \fB\-topics\fR''
will show you what topics exist and
the number of messages in each.
Topics selected are marked with asterisks.
\&``\fINews \fB\-help\fR''
is like
\&``\fInews \fB\-topics\fR''
except that it also displays the syntax of the
.I news
command.
.PP
\&``\fINews \fB\-check\fR''
will report the existence of unread news in the selected topics.
.I News
without any option arguments
will display all unread news items from
all selected topics.
\&``\fINews \fB\-display\fR''
is identical to
.I news
without an argument, but it suppresses the
no news message (useful in login scripts).
\&``\fINews \fB\-review\fR [#] [topic]''
will redisplay the last # (default all) items
in the selected topics.
.PP
News stores in your MH profile the message number of the most recent news
item it has shown you in each topic.
\&``\fINews \fB\-update\fR''
will update your MH profile to indicate no
outstanding news items to be read.
.PP
\&``\fINews \fB\-send\fR topic''
is identical to
\&``mail news.topic''.
The mail sent to news.topic is
automatically filed in a folder with
name topic.
.PP
Any unrecognized switches to
.I news
are sent along to
.IR l (1)
for displaying items, or to
.IR mail (1)
for sending a news item.
.SH FILES
.ta 2.4i
$HOME/.mh\(ruprofile	The user profile
.br
~news/<topics>	The news folders
.br
~news/.<topics>	The number of items in <topics>
.SH PROFILE COMPONENTS
News-topics:            Topics the user is interested in.
.br
News-<topic>:           Current message in news topics
.SH SEE ALSO
mh-mail(5), mh(1)
.SH DEFAULTS
The topic `everyone' is the default topic.
.SH CONTEXT
The current message in each news folder is maintained in the
user's mh\(ruprofile.


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

docs/historical/mh-jun-1982/progs/news.c
#ifdef COMMENT
	Proprietary Rand Corporation, 1981.
	Further distribution of this software
	subject to the terms of the Rand
	license agreement.
#endif

#include "../mh.h"
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/dir.h>
#include <strings.h>

extern char *sprintf();

#define max(a,b) (a > b ? a : b)
#define YES 1
#define NO  0

#define NEWSP   "/usr/news"     /* MUST be same as login News's login dir */
#define MAXTOPICS       50      /* Enough for a while */

int     vecp, topicp;
int     topicspec;      /* topic argument(s) given */
char    *topics[MAXTOPICS+1];
char    *vec[MAXARGS];
int     fdisplay, fcheck, fupdate, fsend, freview, fbody;  /* flags */
int     frevback, fverbose, fhelp, ftopics; /* flags */
int     hit;

struct nts {
	char    t_name[16];
	int     t_num;
} nts[MAXTOPICS+1], *check();
struct nts *ntps[MAXTOPICS];
int nnts;       /* number of actual topics */

struct swit switches[] = {
	"add",          -1,     /* 0 */
	"body",         -1,     /* 1 */
	"check",        0,      /* 2 */
	"display",      0,      /* 3 */
	"review [#]",   0,      /* 4 */
	"send topic ...",0,     /* 5 */
	"topics",       0,      /* 6 */
	"update",       0,      /* 7 */
	"help",         4,      /* 8 */
	0,              0,
};

extern  char _sobuf[];          /* MLW  standard out buffer */

/*ARGSUSED*/
main(argc, argv)
	int argc;
	char *argv[];
{
	register int i;
	register char *cp, **ap;
	char *arguments[50], **argp;

	invo_name = argv[0];
	setbuf(stdout, _sobuf);
#ifdef NEWS
	m_news();
#endif
	if (chdir (NEWSP) < 0) {        /* <-- N.B. */
		fprintf(stderr, "Can't change directory to ");
		perror(NEWSP);
		done(1);
	}
	vecp = 2;
	cp = r1bindex(argv[0], '/');
	if((cp = m_find(cp)) != NULL) {
		ap = brkstring(cp = getcpy(cp), " ", "\n");
		ap = copyip(ap, arguments);
	} else
		ap = arguments;
	VOID copyip(argv+1, ap);
	argp = arguments;
	while(cp = *argp++) {
		if(*cp == '-')
			switch(smatch(++cp, switches)) {
			case -2:ambigsw(cp, switches);       /* ambiguous */
				done(0);
							     /* unknown */
			case -1:vec[vecp++] = --cp; continue;
			case 0:
			case 5: fsend = YES; continue;
			case 1: fbody = YES;
				vec[vecp++] = --cp; continue;
			case 2: fcheck = YES; continue;
			case 3: fdisplay = YES; continue;
			case 4: freview = YES;
				if(**argp >= '0' && **argp <= '9')
					frevback = atoi(*argp++);
				continue;
			case 7: fupdate = YES; continue;

			case 8:
				fhelp = YES;
			case 6:
				ftopics = YES;
				continue;
			}
		else
			if(vecp == 2) {
				VOID addtopic (cp, YES);
				topicspec = YES;
			}
			else
				vec[vecp++] = cp;
	}
	if (!topicspec)
		/* user didn't specify topics so give him his defaults */
		gettopics();
	getnts();
	if (fhelp) {
		help(
"news [topic ...] [switches] [switches for \"c\" or \"mail\" ]", switches);
		putchar('\n');
	}
	if (ftopics) {
		showtopics();
		done(0);
	}
	if(fsend) {
		if(!topicspec) {
			fprintf(stderr, "Usage: news -send topic [mail switches]\n");
			done(1);
		}
		for(i = 0; i < topicp; i++)
			if(check(topics[i]))
				send(topics[i]);
			else
				printf("Unknown topic: %s\n", topics[i]);
		done(0);
	}
	for(i = 0; i < topicp; i++)
		if(check(topics[i]) == NULL)
			printf("Topic: %s unknown.\n", topics[i]);
		else
			disp(topics[i], fverbose || topicspec);
	if(fcheck) {
		if(hit)
			printf(".\n");
	} else if(!hit && !topicspec && !fupdate && !fdisplay)
		nonews();
	m_update();
	done(0);
}


struct nts *
check(topic)
	char *topic;
{
	register struct nts *t;

	for(t = nts; t->t_name[0]; t++)
		if(strncmp(topic, t->t_name, DIRSIZ) == 0)
			return t;
	return 0;
}


disp(topic, argflg)
	register char *topic;
	int argflg;
{
	register struct nts *t;
	register char *cp, *np;
	register int msgnum;
	int high;
	char buf[128];

	if((t = check(topic)) == 0)
		fprintf(stderr, "HUH?\n");
	if((cp = m_find(np = concat("news-", topic, NULLCP))) == NULL)
		cp = "0";
	high = atoi(cp);
	if(fcheck) {
		if(t->t_num > high) {
			if(!hit++)
				printf("Unread news in");
			if(hit > 1)
				printf(",");
			printf(" '%s'", topic);
		}
		return;
	}
	if(fupdate) {
		if(t->t_num > high) {
			m_replace(np, getcpy(m_name(t->t_num)));
			printf("Skipping %d items in %s.\n",
				t->t_num - high, topic);
		}
		return;
	}
	if(freview)
		if(frevback)
			msgnum = max(high - frevback, 0);
		else
			msgnum = 0;
	else
		msgnum = high;
/***    msgnum = freview? frevback? high - frevback : 0 : high; */
	if(msgnum >= t->t_num) {
		if(argflg)
			printf("%s: no new news.\n", topic);
		return;
	}
	VOID sprintf(buf, "%s/%s", NEWSP, topic);
	if(chdir(buf) == -1) {
		perror(buf);
		return;
	}
	vec[1] = showproc;
	for( ; msgnum < t->t_num;) {
		cp = m_name(++msgnum);
		if(hit) {
			printf("\nPress <return> for %s:%s...", topic, cp);
			VOID fflush(stdout);
			VOID read(0, buf, sizeof buf);
		} else
			printf("News item %s:%s\n", topic, cp);
		if(msgnum > high) {
			m_replace(np, getcpy(cp));
			m_update();
		}
		putenv("mhfolder", topic);
		vec[vecp] = cp;
		putchar('\n');
		call(vec + 1);
		hit = 1;
	}
}


call(vector)
	char **vector;
{
	register int pid, child;
	int status;

	VOID fflush(stdout);
	while((child = fork()) == -1) {
		printf("No forks...\n"); VOID fflush(stdout);
		sleep(2);
	}
	if(child == 0) {
		execv(vector[0], vector);
		perror(vector[0]);
		done(1);
	}
	while((pid = wait(&status)) != -1 && pid != child) ;
	if(pid == -1 || status) {
		fprintf(stderr, "Abnormal termination from %s\n", vector[0]);
		done(1);
	}
}


send(topic)
	register char *topic;
{
	vec[0] = mailproc;
	vec[1] = concat("news.", topic, NULLCP);
	if(!fbody)
		printf("Enter text for %s\n", topic);
	call(vec);
	free(vec[1]);
}


#ifdef CUTE
char    *nons[] = {
	"No new news",
	"No news to peruse",
	"Only old news",
	"News shortage",
	"News reporters on strike",
	"Report your own news",
	"News presses broken"
};
#define NONS    (sizeof nons/ sizeof nons[0])
#endif

nonews()
{
#ifdef CUTE
#include <sys/timeb.h>
	struct timeb tb;

	ftime(&tb);
	printf("%s.\n", nons[tb.millitm % NONS]);
#else
	printf("No new news.\n");
#endif
}

showtopics()
{
	register int i;

	sortnts();
	printf("Chk Items Topics\n");
	printf("--- ----- ------\n");
	for(i = 0; i < nnts; i++)
		printf(" %s  %4d  %s\n",
		       addtopic (ntps[i]->t_name, NO) ? "*" : " ",
		       ntps[i]->t_num, ntps[i]->t_name);
	return;
}

gettopics ()
{
	register char **ap;
	register char *cp;

	VOID addtopic ("everyone", YES);
	if((cp = m_find("news-topics")) != NULL) {
		for (ap = brkstring (cp = getcpy(cp), " ", "\n")
		    ; *ap; ap++)
			VOID addtopic (*ap, YES);
	}
	return;
}

addtopic (cp, real)
char *cp;
int real;
{
	register char **cpp;

	for (cpp = topics; *cpp; cpp++)
		if (!strcmp (*cpp, cp))
			return YES; /* We have it already */
	if (real) {
		if (!strcmp ("*", cp)) {
			register int i;
			getnts ();
			for(i = 0; i < nnts; i++)
				VOID addtopic (nts[i].t_name, YES);
		}
		else if (topicp < MAXTOPICS - 1)
			topics[topicp++] = cp;
	}
	return NO;     /* We don't have it already */
}

getnts()
{
	/* shouldn't ALWAYS look up all news folders.
	 * Should only do that for a -topics or -help.
	 */
	struct direct dir;
	struct stat st;
	register struct nts *t;
	register FILE *d;
	char tbuf[DIRSIZ + 2];

	if (nnts > 0)
		/* we've done it already */
		return;
	t = nts;
	if((d = fopen(".", "r")) == NULL) {
		fprintf(stderr, "Can't open ");
		perror(NEWSP);
		done(1);
	}
	tbuf[0] = '.';
	while(fread((char *)&dir, sizeof dir, 1, d)) {
		if(dir.d_ino && dir.d_name[0] != '.') {
			strncpy(t->t_name, dir.d_name, DIRSIZ);
			strcpy(&tbuf[1], t->t_name);
			if(stat(tbuf, &st) != -1)
				t->t_num = st.st_size;
			nnts++;
			t++;
		}
	}
	VOID fclose(d);
	return;
}

sortnts()
{
	register int i;
	extern int ntcmp();

	for (i = 0; i < nnts; i++)
		ntps[i] = &nts[i];
	qsort ((char *) ntps, nnts, sizeof ntps[0], ntcmp);
	return;
}

ntcmp (n1, n2)
struct nts **n1;
struct nts **n2;
{
	return strcmp ((*n1)->t_name, (*n2)->t_name);
}
